tests/bsim/bluetooth/mesh/src/test_proxy_sol.c

```diff
@@ -0,0 +1,304 @@
+	.addr = 0x0002,
+	.dev_key = {0x02},
+};
+
+static struct bt_mesh_model models[] = {
+	BT_MESH_MODEL_CFG_SRV,
+	BT_MESH_MODEL_CFG_CLI(&cfg_cli),
+	BT_MESH_MODEL_PRIV_BEACON_SRV,
+	BT_MESH_MODEL_PRIV_BEACON_CLI(&priv_beacon_cli),
+	BT_MESH_MODEL_OD_PRIV_PROXY_SRV,
+};
+
+static struct bt_mesh_elem elems[] = {
+	BT_MESH_ELEM(0, models, BT_MESH_MODEL_NONE),
+};
+
+static const struct bt_mesh_comp comp = {
+	.cid = TEST_VND_COMPANY_ID,
+	.vid = 0xaaaa,
+	.pid = 0xbbbb,
+	.elem = elems,
+	.elem_count = ARRAY_SIZE(elems),
+};
+
+static bool is_tester_address(void)
+{
+	return bt_mesh_primary_addr() == tester_cfg.addr;
+}
+
+static uint8_t proxy_adv_type_get(uint8_t adv_type, struct net_buf_simple *buf)
+{
+	uint8_t type;
+	uint8_t len = buf->len;
+
+	if (adv_type != BT_GAP_ADV_TYPE_ADV_IND || len < 12) {
+		return 0xFF;
+	}
+
+	(void)net_buf_simple_pull_mem(buf, 11);
+	type = net_buf_simple_pull_u8(buf);
+
+	if (len != ((type == BEACON_TYPE_NET_ID) ? BEACON_NET_ID_LEN : OTHER_ADV_TYPES_LEN)) {
+		return 0xFF;
+	}
+
+	return type;
+}
+
+static void scan_recv(const struct bt_le_scan_recv_info *info, struct net_buf_simple *ad)
+{
+	uint8_t type = proxy_adv_type_get(info->adv_type, ad);
+
+	if (is_tester_address() && type == BEACON_TYPE_PRIVATE_NET_ID) {
+		k_sem_give(&beacon_sem);
+	}
+}
+
+static struct bt_le_scan_cb scan_cb = {
+	.recv = scan_recv,
+};
+
+static void tester_configure(void)
+{
+	uint8_t err, status;
+
+	k_sem_init(&beacon_sem, 0, 1);
+	bt_le_scan_cb_register(&scan_cb);
+
+	bt_mesh_test_cfg_set(NULL, WAIT_TIME);
+	bt_mesh_device_setup(&prov, &comp);
+
+	ASSERT_OK_MSG(bt_mesh_provision(test_net_key, 0, 0, 0, tester_cfg.addr, tester_cfg.dev_key),
+		      "Node failed to provision");
+
+	err = bt_mesh_cfg_cli_app_key_add(0, tester_cfg.addr, 0, 0, test_app_key, &status);
+	if (err || status) {
+		FAIL("AppKey add failed (err %d, status %u)", err, status);
+	}
+}
+
+static void iut_configure(void)
+{
+	uint8_t err, status;
+
+	bt_mesh_test_cfg_set(NULL, WAIT_TIME);
+	bt_mesh_device_setup(&prov, &comp);
+
+	/* Configurations stored in flash during immediate_replay_attack */
+	if (!bt_mesh_is_provisioned()) {
+		ASSERT_OK_MSG(
+			bt_mesh_provision(test_net_key, 0, 0, 0, iut_cfg.addr, iut_cfg.dev_key),
+			"Node failed to provision");
+
+		err = bt_mesh_cfg_cli_app_key_add(0, iut_cfg.addr, 0, 0, test_app_key, &status);
+		if (err || status) {
+			FAIL("AppKey add failed (err %d, status %u)", err, status);
+		}
+	}
+
+	err = bt_mesh_cfg_cli_gatt_proxy_set(0, iut_cfg.addr, BT_MESH_GATT_PROXY_DISABLED, &status);
+	if (err || status) {
+		FAIL("Proxy state disable failed (err %d, status %u)", err, status);
+	}
+
+	err = bt_mesh_priv_beacon_cli_gatt_proxy_set(0, iut_cfg.addr, BT_MESH_GATT_PROXY_DISABLED,
+						     &status);
+	if (err || status) {
+		FAIL("Private proxy state disable failed (err %d, status %u)", err, status);
+	}
+
+	ASSERT_OK_MSG(bt_mesh_od_priv_proxy_set(1), "On-Demand Private Proxy enable failed");
```

Comment:
It is better to use the client model to toggle this state. Yes, it is a public API and we can use it, but doing this through the model we also test the models.